pkg/auth/providers/azure/clients/ad_graph_client.go

```diff
@@ -0,0 +1,237 @@
+package clients
+
+import (
+	"context"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type azureADGraphClient struct {
+	servicePrincipal *adal.ServicePrincipalToken
+	userClient       graphrbac.UsersClient
+	groupClient      graphrbac.GroupsClient
+}
+
+func (c azureADGraphClient) GetUser(id string) (v3.Principal, error) {
+	user, err := c.userClient.Get(context.Background(), id)
+	if err != nil {
+		return v3.Principal{}, err
+	}
+	return UserToPrincipal(user)
+}
+
+func (c azureADGraphClient) ListUsers(filter string) ([]v3.Principal, error) {
+	users, err := c.userClient.List(context.Background(), filter, "")
+	if err != nil {
+		return nil, err
+	}
+
+	var principals []v3.Principal
+	for _, u := range users.Values() {
+		principal, err := UserToPrincipal(u)
+		if err != nil {
+			return nil, err
+		}
+		principals = append(principals, principal)
+	}
+	return principals, err
+}
+
+func (c azureADGraphClient) GetGroup(id string) (v3.Principal, error) {
+	group, err := c.groupClient.Get(context.Background(), id)
+	if err != nil {
+		return v3.Principal{}, err
+	}
+	return GroupToPrincipal(group)
+}
+
+func (c azureADGraphClient) ListGroups(filter string) ([]v3.Principal, error) {
+	groups, err := c.groupClient.List(context.Background(), filter)
+	if err != nil {
+		return nil, err
+	}
+
+	var principals []v3.Principal
+	for _, u := range groups.Values() {
+		principal, err := GroupToPrincipal(u)
+		if err != nil {
+			return nil, err
+		}
+		principals = append(principals, principal)
+	}
+	return principals, err
+}
+
+func (c azureADGraphClient) ListGroupMemberships(id string) ([]string, error) {
+	securityEnabledOnly := false
+	params := graphrbac.UserGetMemberGroupsParameters{
+		SecurityEnabledOnly: &securityEnabledOnly,
+	}
+	groups, err := c.userClient.GetMemberGroups(context.Background(), id, params)
+	if err != nil {
+		return nil, err
+	}
+	return *groups.Value, nil
+}
+
+func (c azureADGraphClient) LoginUser(credential *v32.AzureADLogin, config *v32.AzureADConfig, _ corev1.SecretInterface) (v3.Principal, []v3.Principal, string, error) {
+	logrus.Debug("[AZURE_PROVIDER] Started token swap with AzureAD")
+
+	azureClient, err := NewClientFromCode(credential.Code, config)
```

Comment:
Good point, I chose the second option. The first is hard to implement, since each new clients takes its own set of arguments, so it's hard to make a single constructor for them.